linera-core/src/local_node.rs

```diff
@@ -0,0 +1,546 @@
+                result = self.handle_certificate(certificate.clone(), blobs).await;
+            }
+            match result {
+                Ok(response) => info = Some(response.info),
+                Err(error) => {
+                    // The certificate is not as expected. Give up.
+                    tracing::warn!("Failed to process network certificate {}: {}", hash, error);
+                    return info;
+                }
+            };
+        }
+        // Done with all certificates.
+        info
+    }
+
+    pub(crate) async fn local_chain_info(
+        &mut self,
+        chain_id: ChainId,
+    ) -> Result<ChainInfo, LocalNodeError> {
+        let query = ChainInfoQuery::new(chain_id);
+        Ok(self.handle_chain_info_query(query).await?.info)
+    }
+
+    pub async fn query_application(
+        &self,
+        chain_id: ChainId,
+        query: &Query,
+    ) -> Result<Response, LocalNodeError> {
+        let mut node = self.node.lock().await;
+        let response = node.state.query_application(chain_id, query).await?;
+        Ok(response)
+    }
+
+    pub async fn describe_application(
+        &self,
+        chain_id: ChainId,
+        application_id: UserApplicationId,
+    ) -> Result<UserApplicationDescription, LocalNodeError> {
+        let mut node = self.node.lock().await;
+        let response = node
+            .state
+            .describe_application(chain_id, application_id)
+            .await?;
+        Ok(response)
+    }
+
+    pub async fn download_certificates<A>(
+        &mut self,
+        mut validators: Vec<(ValidatorName, A)>,
+        chain_id: ChainId,
+        target_next_block_height: BlockHeight,
+    ) -> Result<ChainInfo, LocalNodeError>
+    where
+        A: ValidatorNode + Send + Sync + 'static + Clone,
+    {
+        // Sequentially try each validator in random order.
+        validators.shuffle(&mut rand::thread_rng());
+        for (name, client) in validators {
+            let info = self.local_chain_info(chain_id).await?;
+            if target_next_block_height <= info.next_block_height {
+                return Ok(info);
+            }
+            self.try_download_certificates_from(
+                name,
+                client,
+                chain_id,
+                info.next_block_height,
+                target_next_block_height,
+            )
+            .await?;
+        }
+        let info = self.local_chain_info(chain_id).await?;
+        if target_next_block_height <= info.next_block_height {
+            Ok(info)
+        } else {
+            Err(LocalNodeError::CannotDownloadCertificates {
+                chain_id,
+                target_next_block_height,
+            })
+        }
+    }
+
+    /// Downloads and stores the specified blobs, unless they are already in the cache or storage.
+    ///
+    /// Does not fail if a blob can't be downloaded; it just gets omitted from the result.
+    pub async fn read_or_download_blobs<A>(
```

Comment:
From the name of the function I would not have guessed that if a blob failed to download it's just going to move on.

In which case maybe we want to change this to either:
1. The name of the function: `optimistically_read_or_download_blobs`
2. Return `Result<(Vec<HashedValue>, Vec<BytecodeLocation>)>` where the second entry in the tuple is the list of failed bytecodes.

Or both?

Reply:
I leave this for @afck to optimize

Reply:
Good point! I'm fine with either approach.